# Assignment 3 

## Short Answer

1. What are objects in Python?

Objects are collections of data and operations which can be performed on the data. These operations are known as methods, or member functions. 



2. What does this mean: argv[0], *args? and why would we use it?

argv[0] - This is the name of the script. argv is a collection of the parameters passed to the script from the command line, the first of which being the name of the script itself.
*args - this allows for an arbitrary number of parameters to be passed to a function. The ability to pass an arbitrary number of parameters may be desirable in the following situations: a function which operates on collections, meaning that each parameter can be assigned to a value in a collection; and when the function calls another function which can take an arbitrary number of parameter.


3. What does len() do?

len() returns the length of a collection. The collection must implement the \_\_len\_\_ magic function.

4. How do you remove a white space from a string?

There are many different ways. It depends on the specific requirements. Several examples are given below:

In [6]:
my_test_string = 'Hello World! This is a test.   '

# To remove all whitespace.
print(my_test_string.replace(" ", ""))

# To remove leading and trailing whitespace
print(my_test_string.strip()) # This is useful when cleaning up user input

# To remove only the first whitespace. This method can
# be used to remove n whitespaces from the string.
import re
print(re.sub(" ", "", my_test_string, count = 1))


HelloWorld!Thisisatest.
Hello World! This is a test.
HelloWorld! This is a test.   


5. Explain split() and sub().

split() - this is a method belonging to the built-in type, String, which returns a list of substrings which are split by a separator character. The default character to split around is the whitespace character. It is also possible to pass the function a second parameter, maxsplit, which defines the maximum number of times the string can be split.

sub() - I am unsure which method to are referring to. re.sub() from the regex module returns a string with the values matched by a regular expression with values provided.

6. How does 'for' loop and 'while' loop differ in Python and when do you choose to use them?

for loop - This type of loop is used to iterate over a collection or across a range. For loops will run the code in the enclosed scope once per item in the collection. The for loop will exit prematurely if a break statement is reached. For loops are best used when each item in a collection need to have the same logic applied to them.

while loop - The while loop will loop as long as the condition is true or until a break statement is reached. While loops are often used to repeat a process until a specific condition is met, for example, this may be used whilst asking for user input, whereby, the program loops until the user enters an appropriate response, at which point a break statement is used to continue with the program. 

7. How to reverse a list?

In [15]:
# Python includes a built-in function which can reverse lists.
myList = [1, 2, 3, 4, 5, 6, 7, 8]
print(f"Forwards: {myList}")

myList.reverse()
print(f"Reversed: {myList}")

# Lists can also be reverse using slicing
myList = [1, 2, 3, 4, 5, 6, 7, 8]
print(f"Forwards: {myList}")
print(f"Reversed: {myList[::-1]}")

# I prefer the use of reverse slicing since calling the reverse function mutates
# the list which may lead to subtle errors which are hard to find.


Forwards: [1, 2, 3, 4, 5, 6, 7, 8]
Reversed: [8, 7, 6, 5, 4, 3, 2, 1]
Forwards: [1, 2, 3, 4, 5, 6, 7, 8]
Reversed: [8, 7, 6, 5, 4, 3, 2, 1]
[1, 2, 3, 4, 5, 6, 7, 8]


8. How do you check if a vale exists in a list?



In [17]:
myList = [1, 2, 3, 4, 5, 6, 7, 8]

# The if in pattern can be used to check to see if a value is in a collection
if 1 in myList:
    print('1 is in myList!')

if 12 in myList:
    print('12 is in myList') # This does not run, since 12 is not in myList

# This is generally the clearest way to do this. 


1 is in myList!


9. Do you already have a favourite library in Python?

I often use Pandas and scipy allot. I suppose they would be my favourite by the amount of use I get out of them. I find that they are both very helpful libraries. I have also enjoyed my limited time using scikit.learn since it provides an easy introduction into machine learning.

10. Reflection on Python practical:

Do you feel comfortable writing a script for a mini-project?

Yes. I feel that I will need to focus on the science of the project more than the scripting since I am relatively confident at coding already, but I am not very confident at genomic analysis.

Are you considering a given project?

I would be interested in doing a project relating to my master's research on *Acinetobacter baumannii*. Perhaps a project relating to phylogenetic analysis of the ATPPRT enzyme, which is encoded in the His operon.




## Programming

1. Write a Python script that checks if genes on the Y or the X have on average more exons (Hint: Use
two functions – one function getting the exon count, one function to calculate the average).

In [3]:
#!/bin/python3
from typing import List

def getExonLengths(chromosomeID: str) -> List[int]:
    
    filePath = "./human-genes.gtf"

    with open(filePath, 'r') as geneFile:

        exonLengths = list()

        for line in geneFile:
            tmp = line.split("\t")

            if tmp[0].lower() == chromosomeID.lower():
                # This strips out the unwanted characters from the string.
                # It may be possible with regex, but that is a headache...
                exonLengths.append(int(tmp[8].split(";")[2][13:-1]))

        return exonLengths

def mean(inList: List[int]) -> float:
    sum = 0
    for val in inList:
        sum += val

    return float(sum) / float(len(inList))

yMeanExonLength = mean(getExonLengths('y'))
xMeanExonLength = mean(getExonLengths('x'))

print(f"Mean exon counts\nX: {xMeanExonLength}\nY: {yMeanExonLength}")


5
23
15
4
5
26
9
13
4
47
15
2
5
12
20
3
17
7
4
14
53
3
15
10
8
8
14
5
17
50
19
13
13
2
38
7
3
39
2
6
16
18
11
3
15
13
10
10
7
32
3
35
2
15
2
18
12
61
12
9
1
17
14
35
5
15
21
1
22
26
18
42
48
69
2
37
6
9
23
1
14
34
30
15
10
9
34
1
3
29
19
27
41
1
3
40
20
34
38
11
16
8
10
17
4
18
2
42
5
20
2
26
19
11
35
1
33
32
9
27
21
22
51
4
24
15
1
61
9
7
16
3
3
5
35
17
13
38
49
18
5
4
3
3
13
10
11
2
19
22
30
18
1
6
1
13
8
13
1
17
17
5
19
24
4
20
6
21
7
18
24
4
1
46
42
39
18
3
24
14
1
30
15
25
41
8
53
41
3
14
4
21
29
14
37
13
4
13
11
11
10
6
40
22
35
11
4
7
1
13
3
44
14
15
3
1
31
109
5
25
32
2
58
8
16
2
4
13
39
1
57
11
25
5
23
17
6
10
5
8
38
41
7
16
2
1
3
6
7
25
20
14
10
6
1
2
28
3
34
1
18
14
48
4
30
3
10
1
8
3
3
2
4
42
2
25
5
16
31
15
6
10
12
29
1
1
9
11
1
8
30
28
5
5
19
3
19
9
8
20
5
2
54
1
1
22
49
5
9
17
20
6
2
1
25
9
24
2
10
24
4
5
16
20
9
2
1
14
38
1
11
1
40
6
11
21
13
14
10
5
6
3
8
2
38
43
9
13
23
5
20
7
34
7
32
22
6
20
41
20
1
10
19
1
4
30
12
15
6
27
25
14
69
7
5
1
32
10
50
15
122
3
18
31
14
22

2. Write a Python script that tests the hypothesis of the scientist that the number of genes is relevant
for the severity of variation in chromosome numbers (Hints: Use dictionary to count genes and sort).
Summarize your finding in 2-3 sentences

In [10]:
#!/usr/bin/python3
# Task: count genes per chromosome
#       print for every chromosome the number of genes starting with the
#       chromosome having the fewest genes and ending with the chromosome having
#       the most genes.

def getGeneCount():

    # Declare the location of the human-genes.gtf file
    filePath = "./human-genes.gtf"

    # Open the file with error handling
    with open(filePath, 'r') as geneFile:

        # Create an empty dictionary to store the chromosome names
        # and number of genes.
        genes = {}

        # Read the geneFile line by line
        for line in geneFile:

            # Split each line by tabs.
            tmp = line.split("\t")
            
            # Check if the key for the given chromosome exists in the dictionary
            if str(tmp[0]) in genes:
                # If the key exists, then increment the count by 1. This works
                # because each line in the file is a single gene.
                genes[str(tmp[0])] += 1
            else:
                # If the key does not exist, create one and set the count to 1.
                genes[str(tmp[0])] = 1

        # Return the dictionary of gene counts to the caller.
        return genes


# Use dictionary comprehension to sort the output of the getGeneCount function.
geneCount = {chromosome: count for chromosome, count in sorted(getGeneCount().items(), key = lambda item: item[1])}

def printGeneCount(countDict):
    for key in countDict:
        print(f"Chromosome {key} has {countDict[key]} genes")

printGeneCount(geneCount)

Chromosome Y has 57 genes
Chromosome 21 has 245 genes
Chromosome 18 has 289 genes
Chromosome 13 has 331 genes
Chromosome 22 has 459 genes
Chromosome 20 has 561 genes
Chromosome 15 has 615 genes
Chromosome 14 has 658 genes
Chromosome 8 has 702 genes
Chromosome 4 has 765 genes
Chromosome 10 has 774 genes
Chromosome 9 has 820 genes
Chromosome X has 837 genes
Chromosome 16 has 882 genes
Chromosome 5 has 895 genes
Chromosome 7 has 918 genes
Chromosome 6 has 1050 genes
Chromosome 12 has 1068 genes
Chromosome 3 has 1075 genes
Chromosome 17 has 1209 genes
Chromosome 2 has 1279 genes
Chromosome 11 has 1313 genes
Chromosome 19 has 1490 genes
Chromosome 1 has 2076 genes



Chromosome 21 has the fewest genes, in the case of females, but not in males due to the presence of the Y chromosome. The Y chromosome has only 57 genes compared to chromosome 21's 245 genes.

Both trisomy 13 and trisomy 18 are not necessarily fatal. Chromosome 18 has the third fewest genes and chromosome 13 has the fourth fewest genes, this supports the hypothesis that severity and gene count are related. 

Three trisomy sex disorders exist, XXX (triple X syndrome), XXY (Klinefelter syndrome), and XYY. None of these are fatal. The sex chromosomes present a special case where the number of genes does necessarily relate to severity of trisomy disorders. 

To conclude, the scientist's hypothesis appears to have merit when the sex chromosomes are not considered. 